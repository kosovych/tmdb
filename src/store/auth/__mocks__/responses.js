export const getUserResponse = {
  data: {
    id: 'id',
    name: 'name',
    avatar: 'avatar',
    username: 'username'
  }
}

export const requestTokenMockResponse = {
  data: {
    request_token: 'request_token'
  }
}

export const requestSessionIDMockResponse = {
  data: {
    session_id: 'session_id'
  }
}

export const loginFailsErrorResponse = {
  response: {
    data: {
      status_message: 'status_message'
    }
  }
}
