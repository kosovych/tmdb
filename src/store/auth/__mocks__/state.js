export const state = {
  auth: {
    loading: false,
    errorMessage: null,
    sessionId: 'sessionId',
    userId: 1,
    displayName: 'User Name',
    username: 'user_name',
    avatarUrl: 'image'
  }
}
