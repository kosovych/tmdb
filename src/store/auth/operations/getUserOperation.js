import { createLogic } from 'redux-logic'
import { get } from 'lodash'

import { ACCOUNT } from 'Constants'
import { GET_USER } from '../types'
import { storeUser } from '../actions'

export const getUserOperation = createLogic({
  type: GET_USER,
  latest: true,
  async process({ action, axios }, dispatch, done) {
    const { sessionId } = action
    const response = await axios.get(ACCOUNT, { params: { session_id: sessionId } })
    const {
      id, name, avatar, username
    } = response.data
    const avatarUrl = get(avatar, ['tmdb', 'avatar_path'])
    dispatch(storeUser({
      userId: id, displayName: name, username, avatarUrl
    }))
    localStorage.setItem('userId', id)
    done()
  }
})
