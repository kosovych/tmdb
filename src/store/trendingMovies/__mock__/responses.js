export const getTrendingMoviesOperationSuccess = {
  data: {
    page: 1,
    results: [
      {
        id: 299536,
        title: 'Avengers: Infinity War'
      }
    ],
    total_pages: 1,
    total_results: 1
  }
}

export const getTrendingMoviesOperationError = {
  message: 'Error'
}
