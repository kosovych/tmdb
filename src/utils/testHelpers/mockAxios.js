export const mockAxios = ({ method, response, reject } = { reject: false, response: null }) => {
  const axios = {
    [method]: () => new Promise((res, rej) => {
      if (reject) {
        rej(response)
      } else {
        res(response)
      }
    })
  }
  axios.spy = jest.spyOn(axios, method)
  return axios
}

export const mockMultiAxios = (mocks) => {
  const httpClient = { spies: {} }
  mocks.forEach((mock) => {
    if (httpClient[mock.method] === undefined) {
      httpClient[mock.method] = jest.fn()
      const spy = jest.spyOn(httpClient, mock.method)
      httpClient.spies[mock.method] = spy
    }
    httpClient[mock.method].mockImplementationOnce(() => new Promise((resolve, deny) => {
      if (mock.reject) {
        deny(mock.response)
      } else {
        resolve(mock.response)
      }
    }))
  })

  return httpClient
}
