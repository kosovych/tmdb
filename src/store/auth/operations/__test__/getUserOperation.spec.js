import { ACCOUNT } from 'Constants'
import { mockAxios } from 'Utils/testHelpers/mockAxios'
import { getUserOperation } from '../getUserOperation'
import { getUserResponse } from '../../__mocks__/responses'

describe('getUserOperation()', () => {
  describe('success', () => {
    const localStorageSpy = jest.spyOn(Storage.prototype, 'setItem')
    const dispatch = jest.fn()
    const done = jest.fn()
    const axios = mockAxios({ method: 'get', response: getUserResponse })
    const sessionId = 'sessionId'
    const action = { sessionId }

    beforeEach(() => {
      jest.clearAllMocks()
    })

    it('has valid attributes', () => {
      expect(getUserOperation).toMatchSnapshot()
    })

    it('should calls right endpoint with right params', async () => {
      await getUserOperation.process({ action, axios }, dispatch, done)
      expect(axios.spy).toHaveBeenCalledWith(ACCOUNT, { params: { session_id: sessionId } })
    })

    it('should calls dispatch', async () => {
      await getUserOperation.process({ action, axios }, dispatch, done)
      expect(dispatch).toHaveBeenCalled()
    })

    it('should calls localStorage.setItem() with right arguments', async () => {
      await getUserOperation.process({ action, axios }, dispatch, done)
      expect(localStorageSpy).toHaveBeenCalledWith('userId', getUserResponse.data.id)
    })
  })
})
