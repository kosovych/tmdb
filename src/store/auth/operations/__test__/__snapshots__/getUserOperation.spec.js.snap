// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getUserOperation() success has valid attributes 1`] = `
Object {
  "cancelType": undefined,
  "debounce": 0,
  "latest": true,
  "name": undefined,
  "process": [Function],
  "processOptions": Object {
    "dispatchMultiple": true,
  },
  "throttle": 0,
  "transform": undefined,
  "type": "auth/GET_USER",
  "validate": undefined,
  "warnTimeout": 60000,
}
`;
